qt/src/SourceManager.cc

```diff
@@ -142,48 +138,67 @@ int SourceManager::addSources(const QStringList& files) {
 	return added;
 }
 
-bool SourceManager::querySourcePassword(const QString& filename, QByteArray& password) const {
-	bool success = true;
-	if(filename.endsWith(".pdf", Qt::CaseInsensitive)) {
-		std::unique_ptr<Poppler::Document> document(Poppler::Document::load(filename));
-		if(document && document->isLocked()) {
-			success = false;
-			bool ok = false;
-			QString message = QString(_("Enter password for file '%1':")).arg(QFileInfo(filename).fileName());
-			QString text;
-			while(true) {
-				text = QInputDialog::getText(MAIN, _("Protected PDF"), message, QLineEdit::Password, text, &ok);
-				if(!ok) {
-					break;
-				}
-				if(!document->unlock(text.toLocal8Bit(), text.toLocal8Bit())) {
-					password = text.toLocal8Bit();
-					success = true;
-					break;
-				}
+bool SourceManager::querySourcePassword(const QString& filename, QByteArray& password, Poppler::Document* document) const {
+	if(document->isLocked()) {
+		bool ok = false;
+		QString message = QString(_("Enter password for file '%1':")).arg(QFileInfo(filename).fileName());
+		QString text;
+		while(true) {
+			text = QInputDialog::getText(MAIN, _("Protected PDF"), message, QLineEdit::Password, text, &ok);
+			if(!ok) {
+				return false;
+			}
+			if(!document->unlock(text.toLocal8Bit(), text.toLocal8Bit())) {
+				password = text.toLocal8Bit();
+				return true;
+			} else {
+				return false;
 			}
 		}
 	}
-	return success;
+	return true;
 }
 
-bool SourceManager::checkTextLayer(const QString& filename) const {
+bool SourceManager::withTextLayer(const Poppler::Document* const document) const {
+	const int pagesNbr = document->numPages();
+
+	for (int i = 0; i < pagesNbr; ++i) {
+		QString text = document->page(i)->text(QRectF());
+		if(!text.isEmpty()) {
+			return true;
+		}
+	}
+	return false;
+}
+
+bool SourceManager::processInputSource(const QString& filename, Source* source, QStringList& filesWithText) const {
 	if(filename.endsWith(".pdf", Qt::CaseInsensitive)) {
-		std::unique_ptr<Poppler::Document> document(Poppler::Document::load(filename));
-		if(document) {
-			const int pagesNbr = document->numPages();
-
-			for (int i = 0; i < pagesNbr; ++i) {
-				QString text = document->page(i)->text(QRectF());
-				if(!text.isEmpty()) {
-					return false;
-				}
-			}
+		Poppler::Document* document = Poppler::Document::load(filename);
```

Comment:
document is never deleted as far as I see

Reply:
Fixed with std::unique_ptr